DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+in certain situations which will be discussed below:
+
+1. If the struct contains any `const`/`immutable` fields, it is impossible to use the copy constructor
+for `opAssign`, because the copy constructor might initialize them. Even if the copy constructor doesn't
+modify the `const`/`immutable` fields the compiler has to analyze the function body to know that, which
+is problematic in situations when the body is missing. In conclusion, `opAssign` can be generated from the
+copy constructor when the `struct` contains only assignable (mutable) fields.
+
+2. The copy constructor signature is : `@implicit this(ref $q1 S rhs) $q2`, where `q1` and `q2` represent the
+qualifiers that can be applied to the function and the parameter (`const`, `immutable`, `shared`, etc.). The
+problem that arises is: depending on the values of `$q1` and `$q2` what should the signature of `opAssign` be?
+A solution might be to generate for every copy constructor its counterpart `opAssign`: `void opAssign(ref $q1 S rhs) $q2`.
+However, when is a `const`/`immutable` `opAssign` needed? There might be obscure cases when that is useful, but those are
+niche situations where the user must step it and clarify what the desired outcome is and define its own `opAssign`. For
+the sake of simplicity, `opAssign` will be generated solely for copy constructors that have a missing `$q2`.
+
+Taking into account the above mentioned restrictions we can generate a single `opAssign` function
+and leverage the existing `opAssign` generation logic for the [[postblit](https://github.com/dlang/dmd/pull/8505)].
+Because `opAssing` is generated only if there are copy constructors that create a mutable object, the copy
+constructor is going to be called (if possible) when the source is passed as a parameter to the `opAssign`
+function.
+
+## Breaking Changes and Deprecations
+
+1. The parameter of the copy constructor is passed by a mutable reference to the
+source object. This means that a call to the copy constructor may legally modify
+the source object:
+
+```d
+struct A
+{
+    int[] a;
+    @implicit this(ref A another)
+    {
+        another.a[2] = 3;
+    }
+}
+
+void main()
+{
+    A a, b;
+    a = b;    // b.a[2] is modified
+}
+```
+
+A solution to this might be to make the reference `const`, but that would make code like
+`this.a = another.a` inside the copy constructor illegal. Of course, this can be solved by means
+of casting : `this.a = cast(int[])another.a`.
+
+2. If `@implicit` is used in existing code for a constructor, the constructor will be silently changed
+to a copy constructor:
+
+```d
+enum implicit = 0;
+struct C
+{
+    @implicit this(ref C another) {}    // normal constructor before DIP, copy constructor after
+}
+```
+
+3. With this DIP `@implicit` becomes a compiler recognised attribute that can be used solely to
+distinguish copy constructors from normal constructors. This will break code that used `@implicits`
+as a used defined attribute:
```

Comment:
I believe @jacob-carlborg has mentioned somewhere that you can avoid this? If that's possible then that'd be a big plus IMO.

Reply:
I honestly think that people should be discouraged to use UDAs that are also special keywords in the compiler.

Reply:
It's possible, that's how the `@selector` attribute is implemented. I don't see a reason making it a keyword.